src/lib/datepicker/calendar.html

```diff
@@ -1,4 +1,7 @@
 <div class="mat-calendar-header">
+
+  <ng-template [cdkPortalOutlet]="calendarHeaderPortal"></ng-template>
```

Comment:
Are you planning to move the `.mat-calendar-controls` below into the `mat-calendar-header`? That way we can allow people to swap out the default controls for a custom set

Reply:
That's a good idea.

In that case it would make sense to add them to `mat-calendar-header` (component `MatCalendarHeader`) which is the default.

So the template of `MatCalendarHeader` should contain:

```html
<div class="mat-calendar-controls">
    <button mat-button class="mat-calendar-period-button"
            (click)="_currentPeriodClicked()"  [attr.aria-label]="_periodButtonLabel">
      {{_periodButtonText}}
      <div class="mat-calendar-arrow" [class.mat-calendar-invert]="_currentView != 'month'"></div>
    </button>

    <div class="mat-calendar-spacer"></div>

    <button mat-icon-button class="mat-calendar-previous-button"
            [disabled]="!_previousEnabled()" (click)="_previousClicked()"
            [attr.aria-label]="_prevButtonLabel">
    </button>

    <button mat-icon-button class="mat-calendar-next-button"
            [disabled]="!_nextEnabled()" (click)="_nextClicked()"
            [attr.aria-label]="_nextButtonLabel">
    </button>
  </div>
```

And then we could remove that code from `calendar.html` since it will be inserted by `<ng-template [cdkPortalOutlet]="_calendarHeaderPortal"></ng-template>`.

However, if someone provides a custom header, she would have to take care of these controls.

Could we make more sub-components for the buttons so they could be combined in a more modular way?

Could we make components for the buttons (previous, next)?  But how would that work with the callback methods that have to be called on `MatCalendar`?

Reply:
Oh sorry, I missed this comment somehow. Yeah, that's pretty much what I was thinking. We _could_ package the buttons up into their own components, but it seems a little bit overkill to me since they're not very complex. We will need to make some of the internal properties and methods on calendar public (remove the `_`) to indicate that  its ok for people to inject the calendar and call those methods in their custom header

Reply:
@mmalerba done in https://github.com/angular/material2/pull/9639/commits/528361e59d84a7207410da12048025691a0caa81

One test is failing in `src/lib/datepicker/calendar.spec.ts`:

```
it('should re-render when the i18n labels have changed',
      inject([MatDatepickerIntl], (intl: MatDatepickerIntl) => {
        const button = fixture.debugElement.nativeElement
            .querySelector('.mat-calendar-period-button');

        intl.switchToMultiYearViewLabel = 'Go to multi-year view?';
        intl.changes.next();
        fixture.detectChanges();

        expect(button.getAttribute('aria-label')).toBe('Go to multi-year view?');
      })
    );
```
And another test does not pass in `src/lib/select/select.spec.ts`:

```
it('should emit to `optionSelectionChanges` when an option is selected', fakeAsync(() => {
        trigger.click();
        fixture.detectChanges();
        flush();

        const spy = jasmine.createSpy('option selection spy');
        const subscription = fixture.componentInstance.select.optionSelectionChanges.subscribe(spy);
        const option = overlayContainerElement.querySelector('mat-option') as HTMLElement;
        option.click();
        fixture.detectChanges();
        flush();

        expect(spy).toHaveBeenCalledWith(jasmine.any(MatOptionSelectionChange));

        subscription.unsubscribe();
      }));
```

I assume that this related to the use of `querySelector` that tries to get en element that was moved into another template (at least in the first case). I do not know why the second test fails. I cannot find the element in the `MatCalendarHeader` template.